FirebaseStorageSwift/Sources/AsyncAwait.swift

```diff
@@ -66,12 +70,33 @@ import FirebaseStorage
     func putFileAsync(from url: URL,
                       metadata: StorageMetadata? = nil) async throws -> StorageMetadata {
       typealias MetadataContinuation = CheckedContinuation<StorageMetadata, Error>
-      return try await withCheckedThrowingContinuation { (continuation: MetadataContinuation) in
-        // TODO: Use task to handle progress and cancellation.
-        _ = self.putFile(from: url, metadata: metadata) { result in
-          continuation.resume(with: result)
+
+      let metadata =
+        try await withCheckedThrowingContinuation { (continuation: MetadataContinuation) in
+
+          if Task.isCancelled {
+            print("xxxxxxxxx")
+            continuation.resume(throwing: StorageError.cancelled)
+            return
+          }
+          let storageTask = self.putFile(from: url, metadata: metadata) { result in
+            print("yyyyy")
+            continuation.resume(with: result)
+          }
+
+          storageTask.observe(StorageTaskStatus.progress) { snapshot in
+            guard let progress = snapshot.progress else {
+              fatalError("Missing progress")
+            }
+            print("\(progress.completedUnitCount) of \(progress.totalUnitCount)")
+            if Task.isCancelled {
+              print("zzzzzzzzzzzzz")
+              storageTask.cancel()
```

Comment:
I can't figure out a test to get into this block.  

Similar issue as https://forums.swift.org/t/how-can-i-get-the-current-task/50677

Reply:
I guess you may wrap the storage task into a parent task (I did not actually tested it myself):
```
let taskHandle = async {
  return try await storage.putFileAsync(...) // The task is created and started.
} 

taskHandle.cancel() // The storage task is being cancelled because the parent task has been cancelled. 
```
See "Unstructured Concurrency" [here](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html) and [cancel()](https://developer.apple.com/documentation/swift/task/3851218-cancel) method docs.

Another approach could be using a [TaskGroup](https://developer.apple.com/documentation/swift/taskgroup) with has [cancelAll](https://developer.apple.com/documentation/swift/taskgroup/3814852-cancelall) method.

Reply:
Thanks, but I'm not sure I follow.  The first suggestion is how the invocation of the `putFileAsync` API is done in the test below.  Here, I'm trying to figure out the implementation.

The challenge I'm having is finding a way to cancel the Objective C task that is running from GTMSessionFetcher. And I can't figure out a way to get a handle to the Swift task to be able to check if it has been cancelled to know that the Objective C task should be cancelled.